### ISSUE TITLE
[css-color-6] `contrast-color()` MVP in Level 5

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#9166]
In the NYC F2F in 2022, we discussed defining an MVP of `contrast-color()` that would only return `black` or `white` depending on what color text is most readable for the background color passed. This would address the most common use case where authors have a dynamic color (often user selected) and need to display text on it. Currently authors need to know a lot about a11y and color theory to calculate this reliably, and more often than not the result is suboptmal. You can see an example in this very page: 
<img width="166" alt="image" src="https://github.com/w3c/csswg-drafts/assets/175836/22d65159-965f-4e0d-b568-a311bcc18570">

However, the only relevant resolution was [this one](https://github.com/w3c/csswg-drafts/issues/7361#issuecomment-1203086677):

> RESOLVED: We have an unspecified default, the best available to the UA, which gets updated, only allowed when there aren't any candidates. If you include any candidates, algorithm is mandatory

But this only defines defaults, not a path for UAs to implement an MVP of this while we settle the details of the extended syntax. For this to happen, the simpler syntax needs to sit at an earlier level of CSS Color. **UAs are *extremely* keen to ship an MVP for this and have been since 2022**, so it's about time we settle on it, without being distracted by the [multitude](https://github.com/w3c/csswg-drafts/issues/7553) of issues for the extended `color-contrast()` syntax. Out of these issues, only these need to be resolved for this MVP, and none of them is really a roadblock:

- Maybe https://github.com/w3c/csswg-drafts/issues/8539 though I’m hard pressed to think of cases where gamut mapping would actually change the result
- Maybe https://github.com/w3c/csswg-drafts/issues/7358 though I’m not sure picking a strategy would change the result here; maybe even ignoring opacity would be fine for this

Apart from those, we'd need to also decide on the syntax, which should be much simpler to do for this than for the extended case, since the point of this is to reduce the number of knobs authors have to understand and tweak and just do the right thing automagically, in line with the [TAG principle on balancing simplicity and complxity](https://www.w3.org/TR/design-principles/#simplicity).

I would propose just `contrast-color(<color>)` in Level 5 (since it's not CR yet) where:
- Contrast algorithm is UA dependent (for background, [this](https://colorjs.io/apps/blackwhite/) is the result with various existing algorithms, though UAs have also been doing [their own explorations](https://web.dev/accent-color/#guaranteeing-contrast) when they implemented `accent-color`). Yes, there is a concern that authors might be unhappy with the text color changing from white to black as UAs refine their algorithms, but remember this would only happen in case it *improves* the result. From conversations with Chrome folks, it appears they are happy to go this route (and have been taking a similar approach for other high level properties, e.g. `text-wrap: pretty`), hopefully other implementors share the same sentiment.
- `<color>` is assumed to be background and the function returns `white` or `black` depending on which one produces the **highest** contrast for text (choosing `white` if there’s a tie). Yes, we have a resolution that it should be mandatory to specify whether the color argument is foreground or background, but in doesn't really make a difference here, so it would be unfortunate if authors had to add this noise everywhere.
- If people are worried about syntax lock-in, we could even use an entirely different function name. E.g. `max-contrast-color(<color>)` or just `max-contrast(<color>)` since for every color, either `white` or `black` produce the maximum contrast. Or name it in a way that indicates it returns a foreground/text color.

**Comment by svgeesus** [comment]
> In the NYC F2F in 2022, we discussed defining an MVP of `contrast-color()` that would only return `black` or `white` depending on what color text is most readable for the background color passed. I thought we had a resolution on this, but I can’t find it.

The resolutions are in [Minutes New York F2F 2022-08-02 Part II: CSS Color](https://lists.w3.org/Archives/Public/www-style/2022Aug/0013.html) and [Minutes New York F2F 2022-08-02 Part III: CSS Color ](https://lists.w3.org/Archives/Public/www-style/2022Aug/0013.html).

I think the one you are thinking of might be

> RESOLVED: We have an unspecified default, the best available to the
              UA, which gets updated, only allowed when there aren't any
              candidates. If you include any candidates, algorithm is
              mandatory (Issue #7361: color-contrast() default
              algorithm)

although (relevant to your proposal above) we also have

> RESOLVED: Whatever keywords for foreground/background are named,
              they are are required (Issue #7359)

**Comment by LeaVerou** [comment]
Thanks @svgeesus, I updated the OP accordingly.

**Comment by stubbornella** [comment]
I’m the PM for Web UI at Chrome, my team would be implementing color contrast if it were ready. I would like to get to resolution on this issue sooner than later because it is a high priority for us this year. This is my suggestion about how we could move forward. We don’t yet know what the optimal algorithm for color contrast will be, and I believe we need to find a way to make this an iterative process, so we can start experimenting with it as soon as possible. 

TL;DR I’m proposing that we specify an MVP contrast-color() function which has a UA controlled algorithm.

We want to support both user needs and author intent in a way that is performant, usable, accessible, readable, and beautiful. An example of where we've done this by initially specifying UA control is  `text-wrap: balance`, where we started with only a couple heuristics, but plan to expand and refine over time.

Looking at this diagram, it is clear that there are a lot of cases where WCAG 2.1 can be improved on, however it may take time to find those opportunities and experiment to determine if they work.

<img width="2039" alt="PastedGraphic-1" src="https://github.com/w3c/csswg-drafts/assets/39398/bd79b390-9f4e-471e-ad93-cfae22a6a303">

If the author specifies the algorithm now, UAs won’t be able to make improvements and up-level the web. However, if we provide a basic color contrast function now, we can always allow more author control over the algorithm in the future. Going the other way around would require each site to manually upgrade, which would take a long time for users to realize the value.

An iterative, UA-first approach would also remove the burden to get everything right up front and instead move us to a more agile process. For example, we could try improvements behind a flag, effectively allowing us to see how that would affect UX across the web. 

Yes, initially this would make the property a bit less predictable for the values in the middle of that chart, but the user benefit of being able to iterate in the UA outweighs author control (just for now). I am excited to see how much better we can make contrast-color() given time and iteration. 

I propose that we: 

1. Specify an MVP contrast-color() function which has a UA controlled algorithm.
1. Follow up on issues against that algorithm that can be considered independently.
1. Begin UA experimentation on algorithm improvements
1. After we’ve discovered the most important levers in contrast from our experiments, consider exposing them to authors on a case by case basis

**Comment by stubbornella** [comment]
To be clearer about what I'm proposing vs what @LeaVerou suggested: 

I don't think we should specify white/black text or anything else about how the algorithm for contrast should work, except that it should always be better than WCAG 2.1 requires. It seems like the existing resolution may already be good enough to get started on that? (I'm not sure, I could definitely be reading it wrong? Especially the candidates part.)

> RESOLVED: We have an **unspecified default**, the best available **to the
> UA**, which gets **updated**

Our goal is maximizing readability and that is a nuanced question as I can see from the _many_ open issues on this. We will undoubtedly want to add more to the algorithm over time, and if we go with white/black, I don't see a path to reducing the contrast later, even if it still meets WCAG 2.1. If we leave it up to the UA (for now), we can experiment towards maximum readability and good contrast without an artificial constraint imposed by any particular existing algorithm, other than the floor, provided by WCAG 2.1.

@LeaVerou What syntax that would leave open the possibility of adding additional arguments to this function in the future? Perhaps a basic function that just takes in a color and returns another. Could we later add additional optional parameters to the function when we know more about where author control could be helpful?

I'd like to take all the ideas that have come through so far, and start experimenting with them in the UA, for the moment with the non-color expert author in mind. I realize that could feel like relinquishing a lot of control to the UA, but what I really mean is that I'd like for us to partner in a tight loop to improve the UA algorithm together. I think the v0 we would ship would look a lot like what @LeaVerou proposed. I'm suggesting we wait to put it into a spec until we feel confident that we've done everything we can in the UA to have good outcomes for users needing contrast, even when sites are built by authors who don't understand color.

**Comment by LeaVerou** [comment]
To summarize for the call, what we need consensus on is:

1. Do we generally agree to ship a very high level `contrast-color()` function focused around calculating a contrasting text color so we can address the most pressing use case while we figure out the specifics for addressing the more complex use cases?
2. What should be UA-dependent and what should be specified for this MVP? The two proposals so far are:
	1. Algorithm is UA-dependent, return value is either `black` or `white` depending on which one produces the maximum contrast as text (@leaverou, @fantasai)
	2. _Everything_ is UA-dependent, including the return value. The only guarantee is that the value returned will be a color that is guaranteed to be readable on the background provided. (@stubbornella)

**Comment by stubbornella** [comment]
> Everything is UA-dependent, including the return value. 

The reason I wanted the text color to also be UA dependent is so we could maximize readability rather than contrast. I've everyone's thoughts in different threads which left me with the impression that readability is ripe for innovation and experimentation.

**Comment by LeaVerou** [comment]
> > Everything is UA-dependent, including the return value.
> 
> The reason I wanted the text color to also be UA dependent is so we could maximize readability rather than contrast. I've everyone's thoughts in different threads which left me with the impression that readability is ripe for innovation and experimentation.

Thanks, I’ve added this point to the summary above.

**Comment by svgeesus** [comment]
I strongly believe that, following the science, the MVP should be restricted to achromatic colors (black or white, in the limiting case). Here is why:

# Lightness and Chroma/Hue: effect on legibility and reading performance

Studies (Legge 1990) of both people with normal color vision, and people with low vision, have shown that **sufficient luminance contrast** is the primary factor affecting reading speed, and that adding chromatic contrast **does not** further increase reading speed. This is true for both those with low vision and those with normal vision:

> We found no advantages of color contrast for low-vision reading. For text composed of 6° characters, all low-vision subjects read better with luminance contrast than with color contrast.

Beretta summarized the state of text readability research
(as of 2009) as follows:

1. Luminance contrast is the dominant factor to address readability
2. The luminance polarity affects readability (negative polarity is more difficult)
3. The luminance of the background affects readability (lighter backgrounds are preferable)
4. An additional color contrast does not facilitate reading when a sufficient luminance contrast exists
5. On achromatic background, wavelengths at the extremes of the spectrum are more difficult for text.

Indeed, Zuffi _et. al._ (2009) found that for more muted color combinations, the presence of chromatic contrast (measured as Δab) _impaired_ reading performance, compared to luminance contrast (measured as ΔL) alone:

> By using regression analysis we verified that in the first case (“Low Chromatic” group) there is a
significant (p-value = 0.003) **inverse** relationship between reading performance and chromatic contrast; execution times increase as chromatic contrast increases, indicating a **penalizing effect** of this parameter. In the second case (“High Chromatic” group), on the contrary, the relationship is not significant, and this is consistent with the conclusions of other studies, which found that the chromatic contrast does not improve reading performance when a sufficient luminance contrast is present.

See their [Figure 3](https://www.semanticscholar.org/paper/Understanding-the-readability-of-colored-text-by-on-Zuffi-Brambilla/d1b45368f98e02362b6d2a38da96f210a199ccab/figure/5)

## References

- Giordano Beretta (2009) _Text color polarity_ [blog](http://www.mostlycolor.ch/2009/05/text-color-polarity.html)

- G E Legge, G S Rubin (1986) __Psychophysics of reading. IV. Wavelength effects in normal and low vision _  J Opt Soc Am A, **3(1)**: pp 40-51. [abstract](https://pubmed.ncbi.nlm.nih.gov/3950791/)

- Gordon E. Legge, David H. Parish, Andrew Luebker, and Lee H. Wurm (1990)  _Psychophysics of reading. XI. Comparing color contrast and luminance contrast_. Journal of the Optical Society of America vol **7** issue 10 pp. 2002-2010 [https://doi.org/10.1364/JOSAA.7.002002](https://doi.org/10.1364/JOSAA.7.002002)

- Silvia Zuffi, Carla Brambilla, Giordano Beretta, Paolo Scala (2009)
_Understanding the readability of colored text by crowd-sourcing on the Web_
[citation](https://www.semanticscholar.org/paper/Understanding-the-readability-of-colored-text-by-on-Zuffi-Brambilla/d1b45368f98e02362b6d2a38da96f210a199ccab) | [fulltext PDF](http://www.hpl.hp.com/techreports/2009/HPL-2009-182.pdf)

**Comment by svgeesus** [comment]
Regarding the transparency issue:

- https://github.com/w3c/csswg-drafts/issues/7358

if the MVP is only returning opaque black or opaque white then there is no problem with text transparency. (Obviously it needs to be taken into account for the general case with a list of candidate colors, which can be partially transparent. But that case is also fairly simple, composite the text color onto the background color and use the result to do the contrast calculation).

**Comment by tabatkins** [comment]
Me, @stubbornella, and a few of our color-related implementors discussed this yesterday.

The plan we'd like to get WG approval for is to, for now, pursue the simplest "output a color that contrasts with this input color" use-case. The other aspects the WG has discussed are important and we want to address them, but we believe we can make valuable incremental progress here without blocking ourselves in the future.

Exact proposal (loosely worded):

-----------

```
contrast-color(<color> max?)
```

Produces a "very light" or "very dark" color, which will contrast well with the input color. The precise algorithm for determining whether to output a light or dark color is UA-defined, as is the precise color produced. However, the output must be within X distance (delta-E units, presumably?) of white or black, and must be opaque. If `max` is specified, the output must be `white` or `black`. The output must make the same "very light" or "very dark" choice whether `max` is specified or not; that is, if it would produce `black` when `max` is specified, then it must produce a "very dark" color when it's not specified, etc.

---------

This definition allows for *some* UA intervention and tweaking over time, producing near-black or near-white grays, or mixing in a little bit of the input hue, if the UA determines that would give the best result. By guaranteeing that the color is nearly black or white, we limit the potential for undesirable clashes (even if you mix in some hue, the maximum possible chroma is drastically limited), and limit the space of possible inputs that could produce a dramatically different result in different UAs, or the same UA over time. We believe this should be sufficient to avoid freezing on a particular algorithm and forcing future us to specify it. 

If the author wants complete control over how they're tweaking their color, `max` gives them assurances of the output, so they can use color-adjust() or color-mix() to very slightly tweak the opacity or mix in a hue of their choice. Or they just get a guaranteed, clean black or white, if that's what they'd prefer.

**Comment by svgeesus** [comment]
![image](https://github.com/w3c/csswg-drafts/assets/2506926/2c499500-82af-4db0-a37c-7b29f36c8384)

**Comment by tabatkins** [comment]
So, in your estimation, which measure should we use for the thresholding, and what value feels reasonable for that threshold?

**Comment by LeaVerou** [comment]
My takeaway from Chris' pad is that things are different for dark and light colors (and they also tend to be different when we're dealing with hues too), so it's hard to define a threshold that won't be too big for some and too small for others 😕

**Comment by svgeesus** [comment]
I don't have a conclusion at this stage, just wanted to get a feel for the color differences with the slightly-off whites and blacks you mentioned on the call.

**Comment by svgeesus** [comment]
If people want to play around with that example [here it is](https://colorjs.io/notebook/?storage=https%3A%2F%2Fgist.github.com%2Fsvgeesus%2F5b8e3327e66248f4294714903fa481be).

**Comment by Myndex** [comment]
# Regarding slightly off white:
Yes, for main content, though there are issues worth discussing for both light and dark mode.

## Light mode first

With a bright background, `#ffffff` can promote eye blistering fatigue, depending on how bright the display is relative to the eye's adaptation state. Considering that many modern devices easily exceed 1000 nits, visual fatigue is an ever-present problem.

One aspect of fatigue results from viewing high luminance relative to adaptation. If it is caused by a greater bleaching of cone opsins than can be replenished, aka the snow blind effect (this is not actual photokeratitis which is exposure to excess UV, but a loss in sensitivity due to high luminance), or from [blue or HEV light](https://www.optometrists.org/general-practice-optometry/guide-to-eye-exams/why-are-annual-eye-exams-important/blue-light-and-screens/computer-eye-strain-from-high-energy-light/#:~:text=Our%20eyes%20absorb%20all%20kinds%20of%20light%2C%20including,your%20computer%20can%20also%20contribute%20to%20eye%20strain.), is a subject of continuing study.

There are some rather unfortunately misunderstandings floating around the internet suggesting that text be light grey, or that you don't want "too much contrast". Most of these ideas are bunk. One in particular seems to be sourced from an 2018 article at UXMovement, which is a source of much false or misleading information.

In light mode, text should ideally be black. Feelings of "too much contrast" are due to too much luminance. This problem led to the following:


### The Paper Reading Experience
Something we've been working on, The Paper Reading Experience, is a design guideline for light mode, intended to emulate the nature of reading black ink on diffuse white paper, but presented on a self-illuminated display.

The problem with just using an off white like `#eeeeee` is that `#eeeeee` will simply become the peak adaptation level within a minute or two, and therefore no different than `#ffffff`.

_Click for full size_
<img alt="Adaptation demo" width="180" src="https://github.com/w3c/csswg-drafts/assets/42009457/a7959f07-b504-43a2-b725-159ccae975ea">

So, to make an "off white" background useful, there needs to be a full white `#ffffff` in the peripheral vision in order to anchor the adaptation state to the peak white of the display.

The [APC-Readability Criterion](https://readtech.org/ARC/) site uses this technique. (note this is a draft, and the responsive aspect of the site is a little buggy on cell phones.)

Background: 80%  Edge: 100%     
<img alt="Paper Reading Experience" width="703" src="https://github.com/w3c/csswg-drafts/assets/42009457/92538695-e33d-4ca3-b0d8-9361ceda1149">



### Earlier Experiments still up
- [A directory page](https://www.myndex.com/WEB/Perception)
- [Color Blind Simulator](https://www.myndex.com/CVD/)
- [This article on SEO](https://www.myndex.com/WEB/RedirectsForSEO)
- [Article on Luminance](https://www.myndex.com/WEB/LuminanceContrast).

Comments are welcome at [Paper Reading Experience Discussion](https://github.com/orgs/A11yReadTech/discussions/8)


## Dark Mode

Unlike light mode, there _is_ a level of "too much contrast" in dark mode because of the darker adaptation resulting in wider pupils, and the nature of glare, scatter particularly of point sources, and especially for those with astigmatism. Halation is more problematic in dark mode than in light mode.

We are evaluating a max contrast for dark mode, testing $L^c\ -90$ for text smaller than 24px, and $L^c\ -85$ for larger/bolder text. Thread to [discuss maximum contrast.](https://github.com/Myndex/SAPC-APCA/discussions/106)

How dark should the background be in dark mode? IMO that is relative to the surrounding environment. At night, in a dark environment, a full black BG could be fine, provided the text is also dimmer. For a `#000000` BG, text is ideally between `#cccccc` and `#e4e4e4`.

A lighter dark mode BG such as `#444444` and `#f4f4f4` might be preferred by some for a lighter environment.

Examples:

![DARK MODE light surround](https://github.com/w3c/csswg-drafts/assets/42009457/e544767d-17c6-4ffd-90d9-89846a684178)


![DARK MODE dark surround](https://github.com/w3c/csswg-drafts/assets/42009457/bfc8fd0a-3e84-4ff9-9d21-39885260bf94)


## Summary
- For SDR displays
    - In light mode, text is ideally full black `#000`, but the background is better as off white, 80%-85%, with peripheral elements at `#fff` to anchor adaptation.
    - In dark mode, text brightness needs to be relative to the background, and the background is ideally relative to the larger surround.
        - Currently testing $L^c\ -85$ to $L^c\ -90$ as maximums for dark mode text.
